### ISSUE TITLE
[Question] Normals type 

### ISSUE THREAD

**Issue by liszto** [xBimTeam/XbimGeometry#102]
Hello guys,
I continue to progress on my Unity prototype using your lib.
I would integrate your normals instead of letting unity recompute them cause it's causing some issue on some meshes extract from your ifc example.

But I was was expecting normals was vertex normals but your seems to be face normals, right ?

**Comment by martin1cerny** [comment]
Normals has to be index specific rather than vertex specific because on mon-curved surfaces there are multiple normals for a single vertex. Cube having a per-vertex normals would appear like a sphere otherwise.

**Comment by liszto** [comment]
Ok I was asking this cause in your Lakeside.ifc file the roof part named : 
'Basic Roof:Generic-150:319720 #501156' get weird rendering in my unity scene. I was thinking it was due to my normals so I tried to compute them myself instead of trust unity automatic recomputing. The result was almsot the same.

But after I checked in your xBim Xplorer and I just noticed that all the normal from this roof element are flipped.
So your ifc viewer apply get rid of back face culling right ? And it renders both side of a mesh ?

**Comment by martin1cerny** [comment]
Yes, that is a general problem. In game industry models are carefully crafted to be simple yet impressive. But BIM models are not created for very different purpose by tools focused on a technical workings of these assets and the resulting models are far from being optimal for rendering. Plus you sometimes get the geometry already triangulated with wrong normals and wrong orientation of edges in polygon loops. So, you can only try to fix these things as much as possible to hide these problems which were created upstream.

**Comment by liszto** [comment]
Ok ! Thanks for this fast answer, btw I tried to open the file "Lakeside.ifc" in unity based on your wexBim format generated from your xBim Geometry project and the normals from the roof seems to be not correctly oriented.
So I tried to open it in your xbim Xplorer and the result is the same which is logic according to me cause the geometry generation follow the same process than your wexBim generation.

So I would compare it in a 3rd software just to be sure, so I use Blender and it's ifc importer and all the normals from the "Lakeside.ifc" file seems to be correctly displayed as you can see in the attached file. 

So my question is : Is it possible xBim.Geometry had an issue during normal computing that could in some case flipped them ?
![clipboard - 13 mars 2018 15-31](https://user-images.githubusercontent.com/408031/37348431-52cf329e-26d4-11e8-81db-77a7e9d42be3.png)

**Comment by shole** [comment]
Yes, this is an issue with xbim meshing
Here's a screenshot of XbimXplorer with backface coloring on;
https://i.imgur.com/CYmd3Er.jpg
Faces are clearly flipped some time in the meshing process.
Here's an issue about the triangle flipping issues;
https://github.com/xBimTeam/XbimGeometry/issues/96

**Comment by liszto** [comment]
Ah thanks i didn't think to check the previous possible issue to find if someone already noticed it. Probably because I started to think it was me.

**Comment by liszto** [comment]
I have to improve my rendering logic into my Unity's project cause currently to render the Lakeside object it takes something like 20-22seconds which is according to me to high.
I will try to reduce this amount of time of 33% (at least)

But right after this, I can start to look at the base code from xBimEssential or xBimGeometry to try to find the issue about this normals flipping which quite annoying currently.

According to @shole it could be a parsing issue so I will start to have a look there ! Except if some people in the xbim team have some idea concerning this issue @martin1cerny ?

**Comment by martin1cerny** [comment]
I'm sure it is not in a parser. We just read the data as it is in the file. But it might be in the geometry engine or in tesselator. There is a code somewhere which tries to fix wrong orientation of polygons in IFC but it might still go wrong.

**Comment by liszto** [comment]
Ok ! I'll take a look as soon as I have enough time ! Thanks !

**Comment by luoyuke** [comment]
@liszto  I also noticed this in unity. You cant simply convert face normal to vertex normal to fit unity's Rendering processor. If one ignores all the Normals and UVs from Xbim, then some meshes will face inwards which leads some invisible instances in Unity. After a quick reseach I came up a lazy solution: force the unity to render both sides of a mesh by writting a costumized shader with forced Cull Off. Bear in mind this will inevitably Impact the overwall Performance.

**Comment by liszto** [comment]
Indeed, it was one of my solution but it's just impossible cause my final goal is to render this on a mobile device. 
First I have to render this on a computer so it's probably quite decent actually but with the mobile parameter double rendering side isn't an option :)

**Comment by luoyuke** [comment]
I've tried some experimental code to assign the xbim normals to all the vertices of that face triangle. 
things got messy, and Array out of bound error thrown everywhere. The Problem is while in Unity Vertex, Normal, UV are in pairs, the faces are not in that perticular order. That's been said, there's no obvious correlation between faces and Vertices. Better to solve this by rewritting the triangleindex order according to the face normals. I have no clue how

**Comment by liszto** [comment]
I'm currently not on the bim side of my project (I must help to finish some clean / fix in our main app).
But as soon as I come back on this project, I'll help myself and you on this problem I think ! Except if you solved it before me ;)

**Comment by luoyuke** [comment]
don't count on me xD, I still have some ergent Tasks to do at Engineering bureau, after that I have to move on and start to developing VR feature with .ifc data. I wont try to solve this issue until it becomes critical in my unity app with oculus rift.

**Comment by Wolgraph** [comment]
Hi.
I propose my solution on here too.
I couldn't fix the problem withing the xbim code, so I made an external class to procedurally check and fix triangles orientation in order them to point outward.
At the moment it is not that efficient. the SimpleHouse.ifc project take some minutes to be created in Unity.
Give a look to the repository I created:
[https://github.com/Wolgraph/GeometryEditor](https://github.com/Wolgraph/GeometryEditor)
